src/kernels/tensorproduct.jl

```diff
@@ -22,11 +22,10 @@ end
 
 Base.length(kernel::TensorProduct) = length(kernel.kernels)
 
+dims_are_compatible(k::TensorProduct, x) = length(k) == length(x)
+
 function (kernel::TensorProduct)(x, y)
-    if !(length(x) == length(y) == length(kernel))
-        throw(DimensionMismatch("number of kernels and number of features
-are not consistent"))
-    end
+    validate_kernel_dims(kernel, x)
```

Comment:
IMO you'd want to call
```suggestion
    validate_kernel_dims(kernel, x, y)
```
here. However, the problem with the way it is implemented in this draft is that everyone has to implement `dim` for the input types they want to use with TensorProduct since `validate_kernel_dims(k, x, y)` calls `validate_dims(x, y)` which just compares `dim(x)` and `dim(y)`. IMO it might be better to use something more generic such as
```julia
dims_are_compatible(x, y) = true
dims_are_compatible(x::AbstractVector, y::AbstractVector) = length(x) == length(y)
```
for `x` and `y` as well and check `dims_are_compatible(x, y)` instead of `dim(x) == dim(y)`. The main reason would be that I think it is better if one can just apply a TensorProduct to arbitrary types and in the worst case it fails downstream than being forced to define `dim(x)`.

Reply:
Sorry, I don't seem to get what is wrong with being forced to define `dim` for new input types. Won't this happen rarely?

Reply:
I work a lot with kernels on spaces of distribution, so it would at least affect me :smile: In general, I just think the package should provide reasonable checks by default (e.g., for `AbstractVector`) but not be in the way of user that just want to use some other input types. `TensorProduct` does not restrict the types of the input arguments, since they are just forwarded to the different kernels, and hence it feels weird to force users to implement `dim` for other types (which they might not even own).